rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,315 @@
+- Some plugins should be aware of other plugins' `loose` option value. For example, we sometimes need to partially transpile optional chaining in the class properties plugin ([babel/babel#11248](https://github.com/babel/babel/pull/11248)). Should we add a new `optionalChainingLoose` option to `proposal-private-methods` and `proposal-class-properties`? Or should we use a cross-plugin communication channel to share it? What about when someone has the private methods plugin enabled, but not the optional chaining one?
+
+This RFC also solves another problem, which is not directly related to the points above:
+- `@babel/preset-env` has a `loose` option, which is forwarded to all the plugins it enables. When a user wants to set `loose` only for a specific feature, they have to manually install the plugin and explicitly add it to their configuration ([babel/babel#6978](https://github.com/babel/babel/issues/6978)).
+
+
+# Detailed design
+
+## New configuration option
+
+This RFC introduces a new top-level option: `assumptions`. This is an object containing different flags to mark different assumptions about the input code as safe and thus allow optimizing them.
+
+```typescript
+type Assumptions = {
+  [assumption: string]: boolean
+};
+```
+
+This new option should be allowed in the following locations:
+- In programmatic options passed, for example, to `babel.transform()` or to `babel-loader`.
+- In `babel.config.json` and `.babelrc.json` top-level options.
+- In `overrides` blocks. This is useful, for example, to enable a possible `noDocumentAll` assumption on a folder containing server-side code and not inside one containing client-side code.
+- In `env` blocks. This is mostly included for completeness, but I can't think about an use case for it: assumptions are about the _input code_, which isn't affected by env variables.
+
+Additionally, `"assumptions"` should also be allowed inside presets.
+There are two main kinds of presets: framework-specific presets, like `babel-preset-react-app`, and company-wide presets, used to share the same plugins accross different projects. Since assumptions describe the input code and not the used plugins, this new option fits better in company-wide presets where the person writing the preset and the person writing the input code have less "degrees of separation", but there are also use cases in framework-specific presets: for example, `babel-preset-react-app` [enables](https://github.com/facebook/create-react-app/blob/c87ab79559e98a5dae2cd0b02477c38ff6113e6a/packages/babel-preset-react-app/create.js#L151-L154) `loose: true` for `@babel/plugin-proposal-class-properties`.
+
+## Configuration merging
+
+Multiple `"assumptions"` objects should be merged using `Object.assign`, and not overwritten like other options.
+This makes it easy, for example, to enable one additional assumption for a specific folder. Also, it still keeps the ability of disabling an assumption simply by setting it to `false`.
```

Comment:
But what it `preset-react` is included in `preset-pedantic`? `preset-pendantic` might just want to expose `preset-react`'s behavior but override it's assumptions.

I think we can handle it as we do with the other boolean options (`retainLines`, `comments`) or non-boolean options (`sourceType`, ...), which is that the last one wins.

Reply:
I am not sure if `retainLines` and `comments` is as popular as `loose` in presets. We should somehow remind preset authors to choose `assumptions` wisely: It may be pretty hard for end users to read through a long list of assumptions.

Reply:
I would be ok with disallowing `: false` in presets initially, and wait for feedback.

Reply:
Added to the RFC text